packages/dashboard-frontend/src/store/Workspaces/index.ts

```diff
@@ -176,10 +183,38 @@ export const actionCreators: ActionCreators = {
   startWorkspace:
     (workspace: Workspace, params?: ResourceQueryParams): AppThunk<KnownAction, Promise<void>> =>
     async (dispatch, getState): Promise<void> => {
+      const openOauthPage = (oauthUrl: string, namespace: string, workspaceName: string) => {
+        try {
+          // build redirect URL
+          const redirectHost = window.location.protocol + '//' + window.location.host;
+          const redirectUrl = new URL('dashboard/w', redirectHost);
+          redirectUrl.searchParams.set(
+            'params',
+            `{"namespace":"${namespace}","workspace":"${workspaceName}"}`,
+          );
+          const oauthUrlTmp = new window.URL(oauthUrl);
+          window.location.href =
+            oauthUrlTmp.toString() + '&redirect_after_login=' + redirectUrl.toString();
+        } catch (e) {
+          throw new Error(`Failed to open authentication page. ${helpers.errors.getMessage(e)}`);
+        }
+      };
       dispatch({ type: 'REQUEST_WORKSPACES' });
       try {
         const state = getState();
         const cheDevworkspaceEnabled = isDevworkspacesEnabled(state.workspacesSettings.settings);
+        if (
+          SessionStorageService.get(SessionStorageKey.AUTHENTICATION_STATUS) !== 'started' &&
+          workspace.devfile.projects
+        ) {
+          SessionStorageService.update(SessionStorageKey.AUTHENTICATION_STATUS, 'started');
+          const project = workspace.devfile.projects.find(p => (p.name = workspace.projects[0]));
+          if (project && project.git) {
+            await WorkspaceClient.restApiClient.refreshFactoryOauthToken(
+              project.git.remotes.origin,
+            );
+          }
+        }
```

Comment:
@vinokurig I would also like to extract these lines into an OAuth service, like https://github.com/eclipse-che/che-dashboard/pull/640/files#r985524100